### PR TITLE
chore(kuma-cp) upgrade go-control-plane

### PR DESCRIPTION
**Pull request by jakubdyszkiewicz**
# Upgrade of Go Control Plane

We are mostly interested in up to date Envoy Protocol Buffers for API V3, but Go Control Plane brings a couple of changes along the way which results in an opportunity to clean some of our code. We had a lot of copy paste code of Go Control Plane with some adjustments. This tech debt results in many difficulties when upgrading the library.

## XDS

In XDS, we had a copy paste of pkg/xds/server/server.go with [fix for cluster warming](https://github.com/kumahq/kuma/pull/331) that was never merged into upstream.
The problem itself was brought up multiple times on the Envoy Slack and issues and the most up to date discussion is [here](https://github.com/envoyproxy/envoy/issues/13009).
I've rewritten the fix into callbacks with a slight change (instead of sending same EDS response twice, I change the version which results in sending EDS).
This is slightly less optimal, but we don't update Clusters that often. Read `resourceWarmingForcer` for more.

The result is that I removed custom server.go and the fix is implemented with callback, which will make upgrades much easier.

## Custom resources in Go Control Plane

Go Control Plane's **server** now supports custom resources.
Just to remind an architecture of the Go Control Plane:
Server requires ConfigWatcher (SnapshotCache). You put a Snapshot (list of resources) into SnapshotCache which triggers sending a response to the client.
The server now supports custom resources, but the regular SnapshotCache still only supports Envoy resources.
Instead, GCP maintainers introduced LinearCache which supports custom resources (of 1 type and you can compose many caches with MuxCache).
The problem with this cache is that it versions every single resource individually and it triggers update on [every single resource change](https://github.com/envoyproxy/go-control-plane/blob/master/pkg/cache/v2/linear.go#L163).
The interface of `LinearCache` is also not convenient for us because of `UpdateResource` and `DeleteResource`. We are rather relying on "replace" operation in our code.

## MADS

MADS is our custom resource that we send using GCP. I leveraged server custom resource support.

The result is that I removed custom pkg/mads/server/mads.go
Our pkg/util/cache.go unfortunately have to stay for now. I'll start a discussion with GCP maintainers about the changes so we can get rid of this. 

## KDS

Similar to the situation above.
Additionally, to make it work I had to introduce a small workaround described in `typeAdjustCallbacks`

The result is that I removed custom pkg/kds/server/kds.go but pkg/util/cache.go stays as is.

## Linter
The linter now loudly complains about deprecation from Envoy protobufs.
I ignored `l.UseOriginalDst` for now as fixing this is out of the scope of this PR.
I removed Zipkin V1 since it does not work anyway anymore with the version that we are using.

## Callbacks
Since now we have so many callbacks, I noticed that we are implementing empty methods with no operation many times, so I introduced NoopCallbacks to ignore the implementation of methods that we don't care about.

## Docs

No docs, internal changes.

### CODE REVIEW

File: pkg/xds/server/resource_warming_forcer.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+//
+//   Updated Cluster is now in warming state until we send DiscoveryResponse with EDS.
+//   We could sent the same DiscoveryResponse again: DiscoveryRequest [type=EDS, version="UUID-2", nonce="2"], but there is no API in go-control-plane to do it.
+//   Instead we set a new version of the Endpoints to force a new EDS exchange:
+//   10) Kuma sends DiscoveryResponse [type=EDS, version="UUID-3", nonce="3"] // triggered because we set snapshot with a new version
+//   11) Envoy sends DiscoveryRequest [type=EDS, version="UUID-3", nonce="3"] // ACK
+//   After this exchange, cluster is now out of the warming state.
+//
+// The same problem is with Listeners and Routes (change of the Listener that uses RDS requires RDS DiscoveryResponse), but since we don't use RDS now, the implementation is for EDS only.
+// More reading of how Envoy is trying to solve it https://github.com/envoyproxy/envoy/issues/13009
```
Comment (nickolaev) [inline pkg/xds/server/resource_warming_forcer.go:56]: Quite an explanation here. Is this a new problem or we just port an old thing to the new world?

**Review by nickolaev — APPROVED** [review]
Obviously a huge simplification of the codebase. I hope the performance will be better too when we measure.
Thanks for the detailed description of the changes in the PR.
One question: as a result of all this, are we getting easier go-control-plane upgrades now?

File: pkg/xds/server/resource_warming_forcer.go (hunk 1 above)
Comment (jakubdyszkiewicz) [reply pkg/xds/server/resource_warming_forcer.go:56]: Old thing ported as a callback. It was coded here pkg/xds/server/server.go (look for peekResponse, scheduleResponse)

**Comment by jakubdyszkiewicz** [comment]
> One question: as a result of all this, are we getting easier go-control-plane upgrades now?

Yes, upgrades should be much easier now. I'm still looking forward to remove `pkg/util/cache.go` so it will be just a regular upgrade, not a special task

File: pkg/kds/server/components.go [hunk 2]
```diff
@@ -34,14 +34,16 @@ func New(log logr.Logger, rt core_runtime.Runtime, providedTypes []model.Resourc
 		return nil, err
 	}
 	callbacks := util_xds.CallbacksChain{
+		&typeAdjustCallbacks{},
+		util_xds.NewControlPlanIdCallbacks(serverID),
```
Comment (lobkovilya) [inline pkg/kds/server/components.go:38]: typo: Plan -> Plane

File: pkg/xds/server/resource_warming_forcer.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	lastEndpointNonce := r.lastEndpointNonces[streamID]
+	r.lastEndpointNonces[streamID] = request.ResponseNonce
+	nodeID := r.nodeIDs[streamID]
+	if nodeID == "" {
+		nodeID = r.hasher.ID(request.Node) // request.Node can be set only on first request therefore we need to save it
+		r.nodeIDs[streamID] = nodeID
+	}
+	r.Unlock()
+
+	if lastEndpointNonce == request.ResponseNonce {
```
Comment (lobkovilya) [inline pkg/xds/server/resource_warming_forcer.go:101]: If I get it right this situation might also happen if CP receives NACK from Envoy. In order to distinguish warming and NACK I guess we have to check `error_detail` as well - https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#nack. What do you think?

Comment (jakubdyszkiewicz) [reply pkg/xds/server/resource_warming_forcer.go:101]: good catch! I'll do extra check on NACK

**Review by lobkovilya — APPROVED** [review]

**Pull request merged**